In [9]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from scipy import stats
data = pd.read_csv('C:/Users/Pyter/DA/BaiTap/house_price_Dống-Da_Hà-Nội_subdata.csv',header=0)
data

area    price  postDate land_certificate house_direction  \
0     50.0  14700.0     44197            So do             NaN   
1     57.0   4200.0     44265            So do             Nam   
2     62.0  11000.0     44268            So do             NaN   
3     26.0   2900.0     44292            So do             NaN   
4     40.0   3950.0     44292            So do             NaN   
..     ...      ...       ...              ...             ...   
666  168.0  23500.0     44325              NaN             NaN   
667   65.0  17500.0     44325            So do             NaN   
668   60.0  17500.0     44197              NaN             NaN   
669   45.0  24800.0     44197              NaN             NaN   
670  107.0  45000.0     44325            So do             NaN   

    balcony_direction  toilet  bedroom  floor        lat        long  \
0                 NaN     4.0      6.0    6.0  21.015698  105.824583   
1                 Nam     3.0      8.0    4.0  21.025208  105.833111   
2            dong-Nam     6.0      5.0    6.0  21.025379  105.801011   
3                 NaN     4.0      3.0    4.0  21.014325  105.834918   
4                 NaN     NaN      NaN    5.0  21.009862  105.838083   
..                ...     ...      ...    ...        ...         ...   
666               NaN     3.0      3.0    3.0  21.000823  105.829097   
667               NaN     NaN      6.0    5.0  21.011495  105.816666   
668               NaN     NaN      NaN    NaN  21.011495  105.816666   
669               NaN     NaN      NaN    NaN  21.016106  105.832157   
670               NaN     NaN      NaN    5.0  21.009123  105.824428   

    property_type  
0       trong ngo  
1       trong ngo  
2       trong ngo  
3       trong ngo  
4       trong ngo  
..            ...  
666       mat pho  
667       mat pho  
668       mat pho  
669       mat pho  
670       mat pho  

[671 rows x 12 columns]

Câu Hỏi:
- Giữa giá nhà và diện tích có tương quan với nhau?
- Giữa giá nhà và tọa độ địa lý (lat, long) có tương quan với nhau
- Giữa thuộc tính land_certificate và property_type có tương quan với nhau
- Hãy  mã hóa lại thuộc tính giá nhà theo đơn vị (nghìn đồng/m2) và sắp xếp giá nhà thành 4 mức tương ứng với các khoảng tứ phân vị. Tiến hành kiểm định tương quan giữa biến giá nhà này với các biến tọa độ vị trí (lat, long)

In [10]:
for index, column in enumerate(data.columns):
    temp = data.loc[:,column].isna()
    if temp.sum() > 0:
        print('Cột', data.columns[index], 'chứa' , temp.sum() , 'giá trị KT')

Cột area chứa 6 giá trị KT
Cột price chứa 54 giá trị KT
Cột land_certificate chứa 302 giá trị KT
Cột house_direction chứa 629 giá trị KT
Cột balcony_direction chứa 647 giá trị KT
Cột toilet chứa 431 giá trị KT
Cột bedroom chứa 351 giá trị KT
Cột floor chứa 305 giá trị KT


In [31]:
data.dropna(subset=['price'],inplace=True)
data.dropna(subset=['area'],inplace=True)
data.dropna(subset=['land_certificate'],inplace=True)

In [32]:
for index, column in enumerate(data.columns):
    temp = data.loc[:,column].isna()
    if temp.sum() > 0:
        print('Cột', data.columns[index], 'chứa' , temp.sum() , 'giá trị KT')

Cột house_direction chứa 323 giá trị KT
Cột balcony_direction chứa 338 giá trị KT
Cột toilet chứa 149 giá trị KT
Cột bedroom chứa 92 giá trị KT
Cột floor chứa 82 giá trị KT


- Giữa giá nhà và diện tích có tương quan với nhau?

In [30]:
r, pvalue = stats.pearsonr(data['area'], data['price'])
print('r=%.3f, pvalue=%.3f' % (r, pvalue))

r=0.229, pvalue=0.000


kết luận: giữa giá nhà và diện tích không tương quan với nhau do r<alpha=0.05 nên ta có thể kết luận là không tương quan

- Giữa giá nhà và tọa độ địa lý (lat, long) có tương quan với nhau

In [39]:
r, pvalue = stats.pearsonr(data['lat'], data['price'])
print('r=%.3f, pvalue=%.3f' % (r, pvalue))

r=-0.096, pvalue=0.070


kết luận: giữa giá nhà và lat không tương quan với nhau do r<alpha=0.05 nên ta có thể kết luận là không tương quan

In [40]:
r, pvalue = stats.pearsonr(data['long'], data['price'])
print('r=%.3f, pvalue=%.3f' % (r, pvalue))

r=0.082, pvalue=0.120


kết luận: giữa giá nhà và long không tương quan với nhau do r<alpha=0.05 nên ta có thể kết luận là không tương quan

- Giữa thuộc tính land_certificate và property_type có tương quan với nhau

In [42]:
contigency= pd.crosstab(data['land_certificate'], data['property_type'])
c, p, dof, expected = stats.chi2_contingency(contigency)
print(p)

1.0


kết luận: giữa land_certificate và property_type không tương quan với nhau do p > alpha=0.05 nên ta có thể kết luận là không tương quan

- Hãy  mã hóa lại thuộc tính giá nhà theo đơn vị (nghìn đồng/m2) và sắp xếp giá nhà thành 4 mức tương ứng với các khoảng tứ phân vị. Tiến hành kiểm định tương quan giữa biến giá nhà này với các biến tọa độ vị trí (lat, long)

In [45]:
data['price_per_sqm'] = data['price'] / data['area']
data['price_category'] = pd.qcut(data['price_per_sqm'], q=4, labels=['Low', 'Medium', 'High', 'Very high'])


In [50]:
from scipy.stats import spearmanr

correlation, p_value = spearmanr(data['price_per_sqm'], data['lat'])

print(f"R: {correlation:.2f}, p-value: {p_value:.4f}")


R: 0.09, p-value: 0.1019


Kết luận: Do p_value:0.1019 với alpha=0.05 nên ta có thể kết luận rằng giữa price_per_sqm và lat không tương

In [51]:
correlation, p_value = spearmanr(data['price_per_sqm'], data['long'])

print(f"R: {correlation:.2f}, p-value: {p_value:.4f}")

R: -0.23, p-value: 0.0000
